plugins/user.py

```diff
@@ -0,0 +1,220 @@
+from will.plugin import WillPlugin
+from will.decorators import respond_to
+from will import settings
+import pyotp
+import qrcode
+import boto
+import boto.s3.connection
+from boto.s3.key import Key
+import pprint
+import cStringIO
+
+def check_user_permission(storage, username, permission):
+    user_permissions = storage.load("user_permissions", {})
+    if len(user_permissions) == 0:
+        if hasattr(settings, "ADMIN_USERS"):
+            for admin_user in settings.ADMIN_USERS.split(','):
+                user_permissions[admin_user] = ['admin', 'grant']
+                storage.say("giving {} admin permissions".format(admin_user))
+            storage.save("user_permissions", user_permissions)
+        else:
+            msg = "No admin users are defined in redis or environment"
+            storage.say(msg)
+    if username in user_permissions:
+        if permission in user_permissions[username]:
+            return 1
+        else:
+            return 0
+    else:
+        return 0
+
+def verify_twofactor(storage, username, token):
+    user_twofactor = storage.load("user_twofactor", {})
+    if username in user_twofactor:
+        totp = pyotp.TOTP(user_twofactor[username])
+        if totp.verify(token):
```

Comment:
It'd be more pythonic to simply `return totp.verify(token)`. Also, since 0 is "falsy" and 1 is "truthy", it's better form to return True/False than 1/0.